src/main/cpp/src/json_utils.cu

```diff
@@ -0,0 +1,236 @@
+  cudf::strings_column_view const& input,
+  rmm::cuda_stream_view stream,
+  rmm::device_async_resource_ref mr)
+{
+  auto const d_input_ptr = cudf::column_device_view::create(input.parent(), stream);
+  auto const default_mr  = rmm::mr::get_current_device_resource();
+
+  // Check if the input rows are either null, equal to `null` string literal, or empty.
+  // This will be used for masking out the input when doing string concatenation.
+  rmm::device_uvector<bool> is_valid_input(input.size(), stream, default_mr);
+
+  // Check if the input rows are either null or empty.
+  // This will be returned to the caller.
+  rmm::device_uvector<bool> is_null_or_empty(input.size(), stream, mr);
+
+  thrust::transform(
+    rmm::exec_policy_nosync(stream),
+    thrust::make_counting_iterator(0),
+    thrust::make_counting_iterator(input.size()),
+    thrust::make_zip_iterator(thrust::make_tuple(is_valid_input.begin(), is_null_or_empty.begin())),
+    [input = *d_input_ptr] __device__(cudf::size_type idx) -> thrust::tuple<bool, bool> {
+      if (input.is_null(idx)) { return {false, false}; }
+
+      auto const d_str = input.element<cudf::string_view>(idx);
+      auto itr         = d_str.data();
+      auto const end   = itr + d_str.size_bytes();
+
+      while (itr < end) {
+        cudf::char_utf8 ch   = 0;
+        auto const chr_width = cudf::strings::detail::to_char_utf8(itr, ch);
+        if (not_whitespace(ch)) { break; }
+        itr += chr_width;
+      }
+
+      bool is_null_literal{false};
+      if (itr + 3 < end &&
+          (*itr == 'n' && *(itr + 1) == 'u' && *(itr + 2) == 'l' && *(itr + 3) == 'l')) {
+        is_null_literal = true;
+        itr += 4;
+      }
+
+      while (itr < end) {
+        cudf::char_utf8 ch   = 0;
+        auto const chr_width = cudf::strings::detail::to_char_utf8(itr, ch);
+        if (not_whitespace(ch)) {
+          is_null_literal = false;
+          break;
+        }
+        itr += chr_width;
+      }
+
+      // The current row contains only `null` string literal and not any other non-empty characters.
+      // Such rows need to be masked out as null when doing concatenation.
+      if (is_null_literal) { return {false, true}; }
+
+      auto const not_eol = itr < end;
+
+      // If the current row is not null or empty, it should start with `{`. Otherwise, we need to
+      // replace it by a null. This is necessary for libcudf's JSON reader to work.
+      // Note that if we want to support ARRAY schema, we need to check for `[` instead.
+      auto constexpr start_character = '{';
+      if (not_eol && *itr != start_character) { return {false, true}; }
+
+      return {not_eol, not_eol};
+    });
+
+  auto constexpr num_levels  = 256;
+  auto constexpr lower_level = std::numeric_limits<char>::min();
+  auto constexpr upper_level = std::numeric_limits<char>::max();
+  auto const num_chars       = input.chars_size(stream);
+
+  rmm::device_uvector<uint32_t> d_histogram(num_levels, stream);
```

Comment:
Have you tested the histogram code with a \n in it? When I did it the delimiter it returned was still \n and I don't know why. I think it would be better to write our own that is just a reduction into a 256-bit value that is a bitmask for which bytes showed up in the data. No need for a count because all we want is a boolean is it there or not value.